// @elastic.io/component-commons-library/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsonSchemaConversionUtil_1 = require("./jsonSchema/jsonSchemaConversionUtil");
const jsonataTransform_1 = require("./jsonataTransform/jsonataTransform");
const logger_1 = require("./logger/logger");
var AttachmentProcessor_1 = require("./attachment/AttachmentProcessor");
exports.AttachmentProcessor = AttachmentProcessor_1.AttachmentProcessor;
var ApiKeyRestClient_1 = require("./authentication/ApiKeyRestClient");
exports.ApiKeyRestClient = ApiKeyRestClient_1.ApiKeyRestClient;
var BasicAuthRestClient_1 = require("./authentication/BasicAuthRestClient");
exports.BasicAuthRestClient = BasicAuthRestClient_1.BasicAuthRestClient;
var NoAuthRestClient_1 = require("./authentication/NoAuthRestClient");
exports.NoAuthRestClient = NoAuthRestClient_1.NoAuthRestClient;
var OAuth2AuthorizationCodeRestClient_1 = require("./authentication/OAuth2AuthorizationCodeRestClient");
exports.OAuth2RestClient = OAuth2AuthorizationCodeRestClient_1.OAuth2RestClient;
var CookieRestClient_1 = require("./authentication/CookieRestClient");
exports.CookieRestClient = CookieRestClient_1.CookieRestClient;
class JsonSchema {
}
JsonSchema.convertJsonSchemaToEioSchema = jsonSchemaConversionUtil_1.convertJsonSchemaToEioSchema;
JsonSchema.makeSchemaInline = jsonSchemaConversionUtil_1.makeSchemaInline;
JsonSchema.convertDotNetTypeToJsonSchemaType = jsonSchemaConversionUtil_1.convertDotNetTypeToJsonSchemaType;
exports.JsonSchema = JsonSchema;
class JsonataTransform {
}
JsonataTransform.jsonataTransform = jsonataTransform_1.jsonataTransform;
exports.JsonataTransform = JsonataTransform;
class Logger {
}
Logger.getLogger = logger_1.getLogger;
exports.Logger = Logger;
